src/ipvs/ip_vs_proto_icmp.c

```diff
@@ -224,8 +225,31 @@ static struct dp_vs_conn *icmp_conn_lookup(struct dp_vs_proto *proto,
         icmp_type = ((struct icmphdr *)ich)->type;
         icmp_code = ((struct icmphdr *)ich)->code;
         if (!is_icmp_reply(icmp_type)) {
-            sport = ((struct icmphdr *)ich)->un.echo.id;
-            dport = icmp_type << 8 | icmp_code;
+            /* support for traceroute (ICMP method) */
+            if(unlikely(icmp_type == ICMP_TIME_EXCEEDED)) {
+                struct iphdr *ip_hdr, _iphdr;
+                /* get inner IP header */
+                ip_hdr = (struct iphdr *)mbuf_header_pointer(mbuf, iph->len +
+                                       sizeof(_icmph), sizeof(_iphdr), &_iphdr);
+ 
+                if(unlikely(!ip_hdr))
+                    return NULL;
+                    
+                if(unlikely(ntohs(ip_hdr->tot_len) < sizeof(_iphdr) + 
+                                                        sizeof(_icmph)))
+                    return NULL;
+ 
+                s_addr.in.s_addr = ip_hdr->daddr;
+                /* get inner ICMP header */
+                ich = ip_hdr + 1;
+                icmp_type = ((struct icmphdr *)ich)->type;
+                icmp_code = ((struct icmphdr *)ich)->code;
+                sport = icmp_type << 8 | icmp_code;
+                dport = ((struct icmphdr *)ich)->un.echo.id;
```

Comment:
Does it work when packet redirect is not ON?

Reply:
> Does it work when packet redirect is not ON?

No, it needs packet redirect. So this feature does no good to performance.